[Lee Care Gene] iP

leecaregene opened the pull request:
# Duke
* Text-based chatbot
* Based on https://github.com/nus-cs2103-AY2223S1/ip
## To run:
1. Download latest binaries [here](https://github.com/leecaregene/ip/releases/tag/Level-10)
2. Run `java -jar duke.jar`
> You may specify the name of the task file by changing `duke_tasks`
```java
public static void main(String[] args) {
    new Duke("duke_tasks").run();
}
```
## To do:
- [x] Managing tasks
- [ ] Error handling
- [ ] Unit tests
- [ ] ~~Taking over the world❤️~~

Commit:
Add Gradle support

leecaregene committed:
Complete Level-1

leecaregene committed:
Complete Level-2

leecaregene committed:
Complete Level-3

leecaregene committed:
Complete Level-4

leecaregene committed:
Complete A-TextUiTesting

leecaregene committed:
Complete Level-5

leecaregene committed:
Complete Level-6

leecaregene committed:
Complete Level-7

leecaregene committed:
Complete Level-8

leecaregene committed:
Merge branch 'branch-Level-7'

leecaregene committed:
Merge branch 'branch-Level-8'

leecaregene committed:
Complete A-MoreOOP

leecaregene committed:
Complete A-Packages

leecaregene committed:
Merge remote-tracking branch 'nus-cs2103-AY2223S1/ip/add-gradle-support'

leecaregene committed:
Complete A-Gradle

leecaregene committed:
Complete A-JUnit

leecaregene committed:
Complete A-JavaDoc

leecaregene committed:
Complete A-CodingStandard

leecaregene committed:
Complete Level-9

leecaregene committed:
Merge branch 'branch-A-JavaDoc'

leecaregene committed:
Merge branch 'branch-A-CodingStandard'

leecaregene committed:
Merge branch 'branch-Level-9'

leecaregene committed:
Add newline to end of files

leecaregene committed:
Complete Level-10

leecaregene committed:
Merge branch 'branch-Level-10'

eugenelmj commented on src/main/java/duke/TaskList.java:
```diff
@@ -0,0 +1,93 @@
+    }
+
+    public List<Task> search(String keyword) {
+        return tasks.stream().filter(x -> x.getDescription().contains(keyword)).collect(Collectors.toList());
```
A really sleek way of performing search! Very apt use of streams. :)

alexandermula replied:
Efficient solution for the search function, using streams is a clever implementation!

eugenelmj commented on src/main/java/duke/MainWindow.java:
```diff
@@ -0,0 +1,52 @@
+
+/**
+ * Controller for MainWindow. Provides the layout for the other controls.
+ */
```
An extra touch of customisation to make the Bot feel more friendly; very nice!

eugenelmj commented on src/main/java/duke/Ui.java:
```diff
@@ -0,0 +1,80 @@
+    public String showResults(List<Task> result) {
+        StringBuilder temp = new StringBuilder("Here are the matching tasks in your list:\n");
+        for (int i = 0; i < result.size(); i++) {
+            temp.append(i + 1);
```
Spacings between operators obeyed. Good.

eugenelmj reviewed (commented):
Overall this was a very pleasant piece of code to look at: methods were well-abstracted, and coding standards were closely adhered to. Nothing to pick on, really. Keep up the good work! 😄

alexandermula commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,57 @@
+     * @return Loaded task list.
+     * @throws DukeException If the task list could not be loaded.
+     */
+    public List<Task> loadTasks() throws DukeException {
+        try {
+            FileInputStream fis = new FileInputStream(path);
+            ObjectInputStream ois = new ObjectInputStream(fis);
+            List<Task> tasks = (List<Task>) ois.readObject();
+            return tasks;
+        } catch (IOException e) {
+            throw new DukeException("An I/O error occurred while reading from '" + path + "'.");
+        } catch (ClassNotFoundException e) {
+            throw new DukeException("Failed to read '" + path + "'.");
```
LGTM, Clean loadTasks function that is easy to understand.

alexandermula commented on src/main/java/duke/Ui.java:
```diff
@@ -0,0 +1,80 @@
+        return in.nextLine();
+    }
+
+    private String prettyPrint(String text) {
+        String[] lines = text.split("\n");
+        StringBuilder temp = new StringBuilder("    ____________________________________________________________\n");
+        for (String line : lines) {
+            temp.append("     ");
+            temp.append(line);
+            temp.append("\n");
+        }
+        temp.append("    ____________________________________________________________\n");
+        return temp.toString();
+    }
```
Good abstraction of styling the outputs with this function.

alexandermula commented on src/test/java/duke/DeadlineTest.java:
```diff
@@ -0,0 +1,28 @@
+    }
+
+    @Test
+    void constructor_wrongFormat_exceptionThrown() {
```
Effective test naming convention.

alexandermula reviewed (commented):
LGTM! Good job!

leecaregene committed:
Complete A-Assertions

Currently assertions are not being used in the project. Additionally,
user input validation is incomplete.

Assertions are needed to ensure correct input validation.

Let's add assertions to the `Parser` class.

Assertions are added first, so that when input validation logic is added
in the future, the assertions can be used for debugging purposes.

leecaregene committed:
Complete A-CodeQuality

Code quality is subpar in the `Parser` and `Ui` classes.

The quality of the code can be improved.

Let's refactor the `Parser` and `Ui` classes.

The `Parser` class is refactored to use a switch statement to improve
the readability of the code compared to a long if else statement.

leecaregene committed:
Merge pull request #1 from leecaregene/branch-A-Assertions

Complete A-Assertions

leecaregene committed:
Merge branch 'master' into branch-A-CodeQuality

# Conflicts:
#	src/main/java/duke/Parser.java

leecaregene committed:
Merge pull request #2 from leecaregene/branch-A-CodeQuality

Merge branch-A-CodeQuality into master

leecaregene committed:
Complete A-CI

leecaregene committed:
Complete B-RecurringTasks

leecaregene committed:
Complete A-BetterGui

leecaregene committed:
Complete A-UserGuide

leecaregene committed:
Update READMEs

Edfernape commented on src/main/java/duke/Main.java:
```diff
@@ -0,0 +1,39 @@
+/**
+ * A GUI for Duke using FXML.
+ */
+public class Main extends Application {
```
very lean and clean code!

Edfernape reviewed (commented):
nicely done
